### ISSUE TITLE
Database structure

### ISSUE THREAD

**Issue by tudurom** [StudieverenigingSTORM/knt#13]
Proposals for how users, products and orders should be represented in the database, basically what data we need to store.

**Comment by tudurom** [comment]
Proposal for the Transaction object.

A Transaction is essentially a change in a user's balance. There are 4 ways to create one:

1. By loading money into the account.
2. By buying something. In this case, the id of the product should be referenced.
3. By direct change by an administrator. In this case, a comment should be provided.
4. Special case, when a user's account is created. Explained below.

It should have fields for the affected user, the transaction type (enum, one of the above 4 cases), the balance delta, the current balance **after** the transaction was applied (this is to quickly compute the current balance at any given time), the referenced product (`NULL` if doesn't apply), and a comment (must not be empty if transaction was initiated by admin).

When an account is first created, a transaction of the 4th type should be made with both the delta and the current balance set to 0.

If this gets in, users don't need a balance field anymore.

**Comment by Nico213** [comment]
So if I understand this right this is very similar to the old (current) system, the downside that they have is that they can never change the price of a product or delete it, so they end up with multiple products with the same name of which some hidden/non-available with different prices. 
Reason, if you change the price of some products this is reflected in the logs such that for old transactions the price now all of a sudden shows  the new price of product. This is not per se a reason to not do it this way (in fact it is how it is done now) but should be kept in mind!

**Comment by tudurom** [comment]
They can change products (not so sure about removing, for the sake of consistency), because the price is reflected in the delta of the transaction.

**Comment by Nico213** [comment]
oh of course! Anyways seems all good to me then!

**Comment by Nico213** [comment]
Regarding Products:
Old system has categories for example 'fris' (soda's) which then can be assigned to products such as pepsi, fanta etc.

I think it could be beneficial for extra oversight in both interfaces since it allows grouping but I am not sure if the extra complexity is worth the additional effort for mvp or ever.

**Comment by tudurom** [comment]
The current system has a product representing all sodas anyway. And the specific soda inventory is done by BorrelCie anyway (using the magic technique of counting), so I don't think we have to implement this (yet??).

**Comment by Nico213** [comment]
Fair enough, I'm just rambling in the hopes at some point I bring up something useful 😅

So in another attempt: 
What will be the restrictions around passwords? 
eg
- Will it be a pin (numerical) or a actual password (if so what charset)
- What range? (min, max)
- Will it be mandatory for a user to have a password or not. If not mandatory how is 'no password'  represented in the db.

We already discussed this somewhat, especially the last point but since that was external to git I figured since its a design choice it's worthwhile to include here briefly
